chain/syncer.go

```diff
@@ -421,13 +345,39 @@ func (syncer *Syncer) HandleNewTipset(ctx context.Context, tsKey types.TipSetKey
 		return nil
 	}
 
-	// Walk the chain given by the input blocks back to a known tipset in
-	// the store. This is the only code that may go to the network to
-	// resolve cids to blocks.
-	chain, err := syncer.collectChain(ctx, tsKey)
-	if err != nil {
-		return err
+	haltOnState := func(t types.TipSetKey) bool {
+		if syncer.chainStore.HasTipSetAndState(ctx, t.String()) {
+			return true
+		}
+		return false
+	}
+
+	var chain []types.TipSet
+	switch syncer.syncMode {
+	case Syncing:
+		chain, err = syncer.fetcher.FetchTipSets(ctx, tsKey, haltOnState)
+		if err != nil {
+			return err
+		}
+		chain = reverse(chain)
```

Comment:
Move this duplicated line after the case block.

You can ease this by changing `exceedsFinalityLimit` to only take a single block rather than a slice of which it only inspects the first element.

Reply:
Created an issue for that here: https://github.com/filecoin-project/go-filecoin/issues/3112
will address in a follow on